urle381b6cba7_IkarosKappler_ikrs_http_tgz-pJ8-ikrs_json_rpc_JSONRPCHandlerJ8/src/ikrs/json/parser/ConfigurableJSONBuilder.java

```diff
@@ -0,0 +1,307 @@
+	this.currentMemberValue = this.lastPoppedValue;
+	return this.lastPoppedValue;
+    }
+
+    //--- BEGIN --------- Override parser event methods ---------- //
+    @Override
+    protected void fireJSONBegin() {	
+    }
+
+    @Override
+    protected void fireJSONEnd() {	
+    }
+
+    @Override
+    protected void fireArrayBegin() {	
+	JSONArray arr = null;
+	if( this.currentMemberName != null )
+	    arr = this.factory.createArray( this.currentMemberName );
+	else if( this.currentArrayIndex != null )
+	    arr = this.factory.createArray( this.currentArrayIndex );
+	else
+	    arr = this.factory.createArray();
+	this.valueStack.push( arr ); // new JSONArray( new LinkedList<JSONValue>() ) );
+	this.memberNameStack.push( this.currentMemberName );  
+	
+	this.currentArrayIndex = new Integer(0);
+	this.arrayIndexStack.push( this.currentArrayIndex );
+    }
+
+    @Override
+    protected void fireArrayElementEnd() {
+	try {
+	    this.valueStack.peek().getArray().add( this.currentMemberValue );		
+	} catch( JSONException e ) {
+	    throw new RuntimeException( "Fatal error: unexpected JSONException caught when trying to retrieve JSON array.", e );
+	}
+    }
+    
+    @Override
+    protected void fireArrayEnd() {
+	this.popFromStack();	
+    }
+
+    @Override
+    protected void fireObjectBegin() {
+	JSONObject obj = null;
+	if( this.currentMemberName != null )
+	    obj = this.factory.createObject( this.currentMemberName );
+	else if( this.currentArrayIndex != null )
+	    obj = this.factory.createObject( this.currentArrayIndex );
+	else
+	    obj = this.factory.createObject();
+
+	this.valueStack.push( obj ); // new JSONObject( new TreeMap<String,JSONValue>() ) );
+	this.memberNameStack.push( this.currentMemberName );
+	this.arrayIndexStack.push( this.currentArrayIndex );
+    }
+
+    @Override
+    protected void fireObjectEnd() {
+	this.popFromStack();	
+    }
+
+    @Override
+    protected void fireMemberBegin() {
+	// NOOP
+    }
+
+    @Override
+    protected void fireMemberNameRead( String token ) {
+	this.currentMemberName = token;
+    }
+
+    @Override
+    protected void fireMemberEnd() {
+	try {
+	    this.valueStack.peek().getObject().put( this.currentMemberName, 
+						    this.currentMemberValue
+						    );
+	} catch( JSONException e ) {
+	    throw new RuntimeException( "Fatal error: unexpected JSONException caught when trying to retrieve JSON object.", e );
+	}	
+    }
+
+    @Override
+    protected void fireNumberRead( String number ) {
+	
+	try {
+
+	    // May throw NumberFormatException
+	    Number num =  JSONNumber.parseNumber(number);
+
+	    if( this.currentMemberName != null )
+		currentMemberValue = this.factory.createNumber( num, this.currentMemberName );
+	    else if( this.currentArrayIndex != null )
+		this.currentMemberValue = this.factory.createNumber( num, this.currentArrayIndex );
+	    else
+		this.currentMemberValue = this.factory.createNumber( num );
+
+	} catch( NumberFormatException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)